### PR TITLE
Add support for "Type T: Task Manged" activities

### LINKED ISSUE #507: Add support for "Type T: Task Manged" activities
This is sub issue of #504 

More detail to follow.

### PR DESCRIPTION
**Pull request by mmoore99**
Fixes #507

This PR implements the following:

- The ability to specify whether to limit the number of volunteers to the number required.  See my comment in #504 for more a detailed description. This capability is in place for both Activities and Tasks.  The "Wait List" capability has not yet been implemented.

- Functionality changes on the Activity Detail page for a  "Task Managed" activity:
  - The "My Assigned Tasks" list was converted to an "accordion" .  Expanding the accordion provides additional information for the task, including: task description, number of users required and available spots, and skills required. The expanded panel also contains buttons for the various actions that can be performed.

  - An additional "Tasks" list, also an accordion, was created that contains all tasks for the activity that the user is not signed up for.  The user can sign up for tasks from this list. A filter option has been provided to hide tasks that are full. If `IsLimitVolunteers` is true for a task, the task name for tasks that are not full is shown in bold.

  - All of the actions, including: volunteering, accepting a task assignment, rejecting a task assignment, marking a task as complete, marking a task as unable to complete and unenrolling from a task have been converted to Ajax calls and no page refreshes occur for these actions.

  - Confirmation messages are now displayed after completion of each action.

- When a user volunteers for a task an `ActivitySignup` is created if one does not already exist for the activity.  when a user unenrolls from a task the `ActivitySignup` is deleted if the user is not signed up for any other tasks in the activity.

screenclip

screenclip 1

screenclip 2

screenclip 3

screenclip 4

screenclip 5

screenclip 6

screenclip 9

### CODE REVIEW

**Comment by tonysurma** [comment]
last one on the stack.  After we know what to do with the others I will do same here.  Can someone also review this as the others have a review already in place?  Thanks.

**Comment by mmoore99** [comment]
### WIP - Changes made during the review process of #518 will need to be reflected and reconciled with this PR.  Please hold off on review of this PR until those changes have been made and this PR has been rebased.

**Comment by tonysurma** [comment]
@mmoore99   please rebase off of master when ready as we now have a fix for the build/test issues!

For reference on rebasing this page helps (replace upstream with HTBox in examples) https://robots.thoughtbot.com/keeping-a-github-fork-updated

**Comment by mmoore99** [comment]
@tonysurma @mheggeseth I believe this PR is now ready for review.  I have updated the screen shots in the initial comment to reflect UI changes that have been previously discussed.  The branch has been rebased off of master.

**Comment by tonysurma** [comment]
Thanks @mmoore99, it looks like you will need (yet another :)) rebase
@mheggeseth @MisterJames can you review

**Comment by tonysurma** [comment]
Thanks @mmoore99, it looks like you will need (yet another :)) rebase
@mheggeseth @MisterJames can you review

**Comment by mmoore99** [comment]
@tonysurma rebase completed

**Comment by mmoore99** [comment]
@tonysurma another rebase completed

**Comment by tonysurma** [comment]
Thx @mmoore99, once we get a pr review I will pull it in.  kindly pinging @MisterJames @mheggeseth @dpaquette :smiley_cat:

**Comment by mheggeseth** [comment]
@tonysurma @mmoore99 I can review (at least some) this weekend. There's a lot here.

**Comment by MisterJames** [comment]
I'll start reviewing this one now @mheggeseth

File: AllReadyApp/Web-App/AllReady/Areas/Admin/Models/TaskSummaryModel.cs [hunk 1]
```diff
@@ -36,5 +37,6 @@ public class TaskSummaryModel
         public List<VolunteerModel> AllVolunteers { get; set; } = new List<VolunteerModel>();
 
         public int AcceptedVolunteerCount => AssignedVolunteers?.Where(v => v.HasVolunteered).Count() ?? 0;
+        public int NumberOfVolunteersSignedUp { get; set; }
```
Comment (MisterJames) [inline AllReadyApp/Web-App/AllReady/Areas/Admin/Models/TaskSummaryModel.cs]: Is this different than `AllVolunteers.Count()`? How is this set/updated differently?

File: AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs [hunk 2]
```diff
@@ -131,33 +148,74 @@ private bool HasTaskSignupEditPermissions(AllReadyTask task)
             }
         }
 
-        [HttpGet]
-        [Route("/Signup/{taskId}")]
-        [Produces("application/json", Type = typeof(TaskViewModel))]
-        public async void Signup(int taskId, string userId = "")
-        {
-            var task = _allReadyDataAccess.GetTask(taskId);
+        //[HttpGet]
```
Comment (MisterJames) [inline AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs]: If this code block is dead, we can delete it.

File: AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs [hunk 3]
```diff
@@ -1,23 +1,40 @@
...
     [Route("api/task")]
     [Produces("application/json")]
     public class TaskApiController : Controller
     {
         private readonly IAllReadyDataAccess _allReadyDataAccess;
+        private readonly IMediator _bus;
+
 
-        public TaskApiController(IAllReadyDataAccess allReadyDataAccess)
+        public TaskApiController(IAllReadyDataAccess allReadyDataAccess, IMediator bus)
```
Comment (MisterJames) [inline AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs]: Just to be consistent...we started by naming `bus` originally, but have switched over to `mediator`. Could you rename this guy?

File: AllReadyApp/Web-App/AllReady/Features/Notifications/NotifyAdminForUserUnenrolls.cs [hunk 4]
```diff
@@ -22,58 +24,93 @@ public NotifyAdminForUserUnenrolls(IMediator mediator, IOptions<GeneralSettings>
...
+                        HtmlMessage = message.ToString(),
+                        EmailRecipients = new List<string> { campaignContact.Contact.Email },
+                        Subject = subject
+                    }
+                };
             await _mediator.SendAsync(command).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                // ignored
```
Comment (MisterJames) [inline AllReadyApp/Web-App/AllReady/Features/Notifications/NotifyAdminForUserUnenrolls.cs]: I see what you're doing here, but this makes it hard to troubleshoot :(

Can you perhaps inject something like an `ILogger<NotifyAdminForUserUnenrolls> logger` into the constructor and log out the exception? Logging to the console is configured for the project, so at least if we had a debugger attached (or we were checking the Azure logs) we'd see the problem with notifications.

**Comment by MisterJames** [comment]
@mmoore99  this was a great PR and you bit off a big chunk of work here. It is much appreciated. I've requested a few small code cleanups, but otherwise:
 - Tests were included
 - DB migrations worked locally for me
 - Unit tests passing locally (and on build server)
 - Application runs
 - From what I believe I understand of the PR and original issue from the request and conversation(s) :eyes: it seems like most of the check boxes were ticked (save the wait list) :)

For future PRs I have the following suggestions: 
 - a couple of the methods in handlers required some DB lookups that were multiple lines (because of the `.ThenInclude`s). It would be great to break out a private method for each of those so that the handler proper is a little easier to read, something like `var thingy = GetThingy(id)`.
 - when you're doing validation in handlers, you could write a `FooValidator` object with public methods that would allow for unit tests on those validations

Finally, I know this was a monumental task, and you broke it out of another monumental task, but I feel like this could have been 4 or 5 seperate PRs. This is a hard thing to ask of a volunteer (especially as a fellow volunteer!) but this one was difficult to evaluate because of the size. Thankfully things were working and your code was at a good level, but when they're this big I can't review them on my lunch hour, if you get what I mean?

Thanks again...I'll merge after your code touches, please ping me.

**Comment by MisterJames** [comment]
@mmoore99 You're also going to have a small rebase to tackle because of #596, but it should only be the one file (ActivityController).

File: AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs (hunk 2 above)
Comment (mmoore99) [inline AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs]: Done.

File: AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs (hunk 3 above)
Comment (mmoore99) [inline AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs]: done

File: AllReadyApp/Web-App/AllReady/Features/Notifications/NotifyAdminForUserUnenrolls.cs (hunk 4 above)
Comment (mmoore99) [inline AllReadyApp/Web-App/AllReady/Features/Notifications/NotifyAdminForUserUnenrolls.cs]: Done.  I agree that is a much better approach.

File: AllReadyApp/Web-App/AllReady/Areas/Admin/Models/TaskSummaryModel.cs (hunk 1 above)
Comment (mmoore99) [inline AllReadyApp/Web-App/AllReady/Areas/Admin/Models/TaskSummaryModel.cs]: `AllVolunteers` refers to volunteers for the activity, not the task. `AssignedVolunteers` refers to volunteers for the task.  At the moment I cannot recall why I did not use `AssignedVolunteers.Count`, so I have changed it to that for now.

**Comment by mmoore99** [comment]
@MisterJames thanks for the review and the comments.  I have made all of the requested changes and rebased.  I fixed a bug that was introduced in the refactoring of `ActivityController` in PR #596.  Everything has now been pushed. My responses to some of your comments follow:

> a couple of the methods in handlers required some DB lookups that were multiple lines (because of the .ThenIncludes). It would be great to break out a private method for each of those so that the handler proper is a little easier to read, something like var thingy = GetThingy(id)

I refactored a number of the handlers as per your suggestion. 

> when you're doing validation in handlers, you could write a FooValidator object with public methods that would allow for unit tests on those validations

I am afraid you lost me here. What validation in handlers are you referring to?  Is there an example that you can point me to that might help clarify?

>  but I feel like this could have been 4 or 5 separate PRs.

Got it!  I apologize for the size of the PR.  I will try to keep them more manageable in the future.

**Comment by MisterJames** [comment]
Awesome, thanks for the quick update!

So, as an example of where an external validator might come in handy is here:

https://github.com/HTBox/allReady/pull/523/files#diff-5370386cd1e40cd809df10a8a13d6a4bR42

Definitely not required at this point; I'm actually still trying to come up with a recommendation there, given that a validator of any capacity would require access to the database, which either means taking a dependency or dispatching via MediatR, so it's more work. I've trying the inline approach in my open PR (#602) right now, but I feel like it's probably best to remove the coupling.

**Pull request merged**